### PR TITLE
Provide some samples or documentation #363

### PR DESCRIPTION
**Pull request by grumpyf0x48**
Include manpages in distribution.zip.

manpage generated by picocli-codegen are added to a manpage folder in distribution.zip.

After build, you can see what it looks like with:

```console
$ man build/docs/manpage/jbang.1
```

i would have liked to zip each file before inclusion in distribution.zip but it looks like i'm too new to Gradle for that yet :-(

### CODE REVIEW

**Comment by maxandersen** [comment]
Not a compute r now so can’t test it but is there html version it can generate ?

**Comment by grumpyf0x48** [comment]
> Not a compute r now so can’t test it but is there html version it can generate ?

For instance i was into adding the manpages.

But yes the picocli generator can generate HTML5 pages as shown in the example.

**Comment by maxandersen** [comment]
Thanks for getting it to work - my attempt failed!

so this seem to generate a man page *per* command/subcommands, 31 man pages in total. That seems excessive and not really user friendly ...like - when I do `man ./jbang.1` I get very little info and no option to navigate/discover the actual docs.

I think a html page published on jbang.dev would be better approach than this.

wdyt ?

**Comment by grumpyf0x48** [comment]
> Thanks for getting it to work - my attempt failed!
> 
> so this seem to generate a man page _per_ command/subcommands, 31 man pages in total. That seems excessive and not really user friendly ...like - when I do `man ./jbang.1` I get very little info and no option to navigate/discover the actual docs.
> 
> I think a html page published on jbang.dev would be better approach than this.
> 
> wdyt ?

Yes the `ManPageGenerator` generates one asciidoc file per 'final' command and then we get one manpage per asciidoc file. I've searched a way of having less files but could not find it so far: i think about asking to Picocli folks about it.

Having an html file with all commands, subcommands and options on jbang.dev would be great, but if we use the same generator, we'll end up with many html pages as well. Maybe we need to find another generator.

**Comment by maxandersen** [comment]
Maybe just create a Uber asciidoc that include all adoc so it All ends up in one?

**Comment by grumpyf0x48** [comment]
> Maybe just create a Uber asciidoc that include all adoc so it All ends up in one?

I've modified the build to have:

- one uber manpage
- one manpage per sub-command

I've used asciidoc templates in the repository (which could probably be generated automatically) and we now end up with 13 manpages files instead of 31.

File: build.gradle [hunk 1]
```diff
@@ -277,6 +283,25 @@ task tag(type: Exec) {
...
+	main 'picocli.codegen.docgen.manpage.ManPageGenerator'
+	args mainClassName, "--outdir=${buildDir}/generated-picocli-docs"
+}
+
+asciidoctor {
+	dependsOn(generateManpageAsciiDoc)
+	sourceDir = file("${projectDir}/misc/doc-templates")
+	outputDir = file("${buildDir}/docs")
+	logDocuments = true
+	backends 'manpage'
```
Comment (maxandersen) [inline build.gradle:302]:
```suggestion
	backends = ['html5', 'manpage']
```

then we also get the html5 version.

File: build.gradle [hunk 2]
```diff
@@ -305,6 +330,7 @@ versioner {
 	}
 }
 
+build.dependsOn(asciidoctor)
```
Comment (maxandersen) [inline build.gradle:333]:
```suggestion
//build.dependsOn(asciidoctor)
```

its quite slow and still not fully releasable thus I would keep `asciidoctor` as an independent task.

File: build.gradle [hunk 3]
```diff
@@ -11,6 +12,7 @@ plugins {
 	id 'de.fuerstenau.buildconfig' version '1.1.8'
 	id "com.dorongold.task-tree" version "1.5"
 	id "com.geoffgranum.gradle-conventional-changelog" version "0.3.1"
+	id "org.asciidoctor.convert" version "2.4.0"
```
Comment (maxandersen) [inline build.gradle:15]: this generates quite the few warnings - tried move it to latest `org.asciidoctor.jvm.convert` but then the whole include setup can't find their files.

File: build.gradle [hunk 4]
```diff
@@ -93,6 +96,9 @@ def commonSpec = project.copySpec {
 	from(project.buildDir) {
 		include 'tmp/version.txt'
 	}
+	from("${buildDir}/docs/manpage") {
```
Comment (maxandersen) [inline build.gradle:99]: I would do the inclusion in a separate PR as it also needs enablement in the various install scripts as otherwise its just "unnecessary bagage"

**Comment by maxandersen** [comment]
> * one uber manpage
> * one manpage per sub-command

this is definitely better.

But it still feels a bit "weak" if you enable html generation you get a nice single page but no links at all which surprised me.

I'm wondering if we are hitting the limits of picocli codegenerator and better if we extended it to generate a bit more useful output ?

File: build.gradle (hunk 2 above)
Comment (grumpyf0x48) [reply build.gradle:333]: ok

File: build.gradle (hunk 1 above)
Comment (grumpyf0x48) [reply build.gradle:302]: I removed html5 because the result was not totally ok: many pages, no links between them.

File: build.gradle (hunk 4 above)
Comment (grumpyf0x48) [reply build.gradle:99]: ok

File: build.gradle (hunk 3 above)
Comment (grumpyf0x48) [reply build.gradle]: it is working now.

**Comment by grumpyf0x48** [comment]
Will we do something with this MR or we close it until Picocli handles properly sub-commands ?

**Comment by maxandersen** [comment]
I think in its current form its not of that great value - I've picked out the asciidcotor setup so that is in master now. Did you see the discussion at https://github.com/remkop/picocli/issues/1227 ?

Thanks for it - I'll close it for now as I i think the generated format and needed maintanence needed of auxilary files are too "weak" to be of real value at this time.

@all-contributors please add @grumpyf0x48  for infra

**Comment by grumpyf0x48** [comment]
> I think in its current form its not of that great value - I've picked out the asciidcotor setup so that is in master now. Did you see the discussion at https://github.com/remkop/picocli/issues/1227 ?
> 
> Thanks for it - I'll close it for now as I i think the generated format and needed maintanence needed of auxilary files are too "weak" to be of real value at this time.
> 
> @all-contributors please add @grumpyf0x48  for infra

I've seen the discussion but nothing of what i have tried for jbang manpages can be reused in  picocli.

To me, even if not perfect, it would have been better than nothing to have the manpages like this, which look very similar to what gît manpages are - without the links between commands.

**Comment by maxandersen** [comment]
for me manpages biggest issue is they are hard to browse - so I really want that solved before committing to maintain the external adoc files.
